cmd/ls.go

```diff
@@ -42,59 +39,163 @@ func cname(entry apitypes.Entry) string {
 	return cname
 }
 
-func formatLSEntries(ls []apitypes.Entry) string {
-	table := make([][]string, len(ls))
-	for i, entry := range ls {
-		var mtimeStr string
-		if entry.Attributes.HasMtime() {
-			mtimeStr = format(entry.Attributes.Mtime())
+// item should be a "file"/"dir" type item. formatItem returns
+// an array of rows representing that item's entries
+func formatItem(item lsItem, longFormat bool) [][]string {
+	var entries []apitypes.Entry
+	if item.Type() != dirItem {
+		// Print the path for "file" items. This is consistent
+		// with the built-in ls
+		item.entry.CName = item.path
+		entries = append(entries, item.entry)
+	} else {
+		entries = item.children
+	}
+
+	var rows [][]string
+	for _, entry := range entries {
+		var row []string
+
+		if !longFormat {
+			row = []string{cname(entry)}
 		} else {
-			mtimeStr = "<unknown>"
+			var mtimeStr string
+			if entry.Attributes.HasMtime() {
+				mtimeStr = formatTime(entry.Attributes.Mtime())
+			} else {
+				mtimeStr = "<unknown>"
+			}
+			verbs := strings.Join(entry.Actions, ", ")
+			row = []string{cname(entry), mtimeStr, verbs}
 		}
 
-		verbs := strings.Join(entry.Actions, ", ")
+		rows = append(rows, row)
+	}
+
+	return rows
+}
 
-		table[i] = []string{cname(entry), mtimeStr, verbs}
+func pad(str string, longFormat bool) []string {
+	if longFormat {
+		return []string{str, "", ""}
 	}
-	return cmdutil.NewTableWithHeaders(headers(), table).Format()
+	return []string{str}
 }
 
 func lsMain(cmd *cobra.Command, args []string) exitCode {
-	// If no path is declared, try to list the current directory/resource
-	path := "."
+	paths := []string{"."}
 	if len(args) > 0 {
-		path = args[0]
+		paths = args
 	}
 	longFormat, err := cmd.Flags().GetBool("long")
 	if err != nil {
 		panic(err.Error())
 	}
 
 	conn := cmdutil.NewClient()
-	entry, err := conn.Info(path)
-	if err != nil {
-		cmdutil.ErrPrintf("%v\n", err)
-		return exitCode{1}
-	}
 
-	var entries []apitypes.Entry
-	if !entry.Supports(plugin.ListAction()) {
-		entries = []apitypes.Entry{entry}
-	} else {
-		children, err := conn.List(path)
-		if err != nil {
-			cmdutil.ErrPrintf("%v\n", err)
-			return exitCode{1}
+	// Fetch the required data
+	var items []lsItem
+	for _, path := range paths {
+		var item lsItem
+		item.path = path
+		item.entry, item.err = conn.Info(path)
+		if item.err == nil && item.Type() == dirItem {
+			item.children, item.err = conn.List(path)
 		}
-		entries = children
+		items = append(items, item)
 	}
 
-	if longFormat {
-		cmdutil.Print(formatLSEntries(entries))
-	} else {
-		for _, entry := range entries {
-			cmdutil.Println(cname(entry))
+	// Sort the items to ensure that the output's
+	// printed in the expected "errors", "files",
+	// and "dirs" order.
+	sort.Sort(lsItemSlice(items))
```

Comment:
If all we need is to sort once, then https://golang.org/pkg/sort/#Slice lets you provide just the less function.

Reply:
Thanks, that's much cleaner.